Support tensor betas in Adam and AdamW

mlazos opened the pull request:
Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134171



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

mlazos committed:
Support tensor betas in Adam and AdamW

[ghstack-poisoned]

awgu commented on torch/optim/adam.py:
```diff
@@ -537,15 +544,29 @@ def _multi_tensor_adam(
 
         torch._foreach_mul_(device_exp_avg_sqs, beta2)
+
+        # Due to the strictness of the _foreach_addcmul API, we can't have a single
```
I wonder if we can add an overload. `scaled_device_grads` being out-of-place is going to use even more intermediates.

mlazos replied:
If we compile these intermediates go away, but agreed that for eager this is a problem and another overload could be a good solution.

mlazos replied:
Well actually now that I think of it perhaps users will only wrap the beta if they're compiling anyway, there isn't really a use case for tensor betas in eager.

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




[ghstack-poisoned]

mlazos commented:
@janeyx99 this is ready for review when you get the chance

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




[ghstack-poisoned]

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




[ghstack-poisoned]

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




[ghstack-poisoned]

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




[ghstack-poisoned]

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




[ghstack-poisoned]

janeyx99 commented:
@mlazos I was assuming this was still in progress as CI is pretty red--do you still want a review now or after you fix CI?

janeyx99 commented on torch/testing/_internal/common_optimizers.py:
```diff
@@ -497,6 +495,16 @@ def optim_inputs_func_adam(device, dtype=None):
+                "capturable": True,
+            },
+            desc="Tensor lr, Tensor betas, with capturable and amsgrad",
+        ),
```
Have not reviewed the rest of PR, but these would be good to address along with CI:
We should also add 3 more variations where:
- capturable is False and amsgrad is False
- capturable is True and amsgrad is False
- capturable is False and amsgrad is True

There should either be a test where only 1 beta is a Tensor if that's supported or a test that it errors nicely if not.

mlazos replied:
SG, I'll add those cases. I'll error nicely, don't think it's worth supporting 1 constant and one tensor

mlazos replied:
Note I also banned the capturable is False case (like we did for LR)

mlazos commented:
> @mlazos I was assuming this was still in progress as CI is pretty red--do you still want a review now or after you fix CI?

Yeah I'll take the review now, there's an issue with torch script and calling `.to` + an import which should fix CI. I'd rather get feedback on the general approach before spending a lot of time debugging if that works

janeyx99 reviewed (commented):
Review on overall approach:
1. Considering the expected e2e user experience for the issue we're trying to solve, how do we expect tensor betas to be set? e.g., we should update docs and constructor checks to allow for Tensor betas. Do we want to error in the LRScheduler that'd modify betas if the betas were not Tensors in compile mode?
2. The `_maybe_copy_beta` function that performs cross-device moves is likely not the way we want to go. Since the betas are most likely going to be CPU scalar tensors, our dispatcher should be able to handle them with CUDA Tensors. Only if the betas are CUDA scalar tensors, then we likely want to maintain a dictionary of betas similar to how we handle lr https://github.com/pytorch/pytorch/blob/main/torch/optim/adamw.py#L653. This is to avoid the perf hit of moving it across device every time.
3. Memory is important to maintain as a minimum for any API that is reachable in eager, for example, through using more inplace ops.

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




[ghstack-poisoned]

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




[ghstack-poisoned]

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




[ghstack-poisoned]

patrick-botco commented:
@janeyx99 @mlazos is this PR still a WIP, anything blocking it from merging? thanks!

mlazos commented:
> @janeyx99 @mlazos is this PR still a WIP, anything blocking it from merging? thanks!

@patrick-botco there appears to be some accuracy issues that I need to look into and I'm on vacation for two more weeks so won't be able to look into it until 28th. Feel free to take over if you'd like, I think the right skeleton is in place to land these last two PRs (although the next one might need some polishing)

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




[ghstack-poisoned]

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




[ghstack-poisoned]

mlazos commented:
> > @janeyx99 @mlazos is this PR still a WIP, anything blocking it from merging? thanks!
> 
> @patrick-botco there appears to be some accuracy issues that I need to look into and I'm on vacation for two more weeks so won't be able to look into it until 28th. Feel free to take over if you'd like, I think the right skeleton is in place to land these last two PRs (although the next one might need some polishing)

@patrick-botco FYI I fixed up the previous PR and it's landing, so I'm now I'm focused on this one. Should get landed in the next week or so if there aren't any new issues uncovered.

patrick-botco commented:
> > > @janeyx99 @mlazos is this PR still a WIP, anything blocking it from merging? thanks!
> > 
> > 
> > @patrick-botco there appears to be some accuracy issues that I need to look into and I'm on vacation for two more weeks so won't be able to look into it until 28th. Feel free to take over if you'd like, I think the right skeleton is in place to land these last two PRs (although the next one might need some polishing)
> 
> @patrick-botco FYI I fixed up the previous PR and it's landing, so I'm now I'm focused on this one. Should get landed in the next week or so if there aren't any new issues uncovered.

thank you!

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




[ghstack-poisoned]

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




[ghstack-poisoned]

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




[ghstack-poisoned]

mlazos commented on torch/optim/adam.py:
```diff
@@ -537,23 +566,37 @@ def _multi_tensor_adam(
         torch._foreach_mul_(device_exp_avg_sqs, beta2)
+
+        # Due to the strictness of the _foreach_addcmul API, we can't have a single
+        # tensor scalar as the scalar arg (only python number is supported there)
```
@janeyx99 lmk what you think of this or if you have any better ideas here, but here's my reasoning for this. I can't use an inplace op here (we can't mutate device_grads) and I also can't pass a scalar tensor as the value argument (This is true for all of the foreach APIs, it can only be number or complex), so I had to add this extra out of place mul.

janeyx99 replied:
Yea, I'm fine with this--could you also open an issue to track since the ideal would be that our foreach APIs can support Tensor values for addcmul

mlazos replied:
Sounds good, I will file an issue.

mlazos replied:
https://github.com/pytorch/pytorch/issues/139795

janeyx99 commented on torch/optim/adam.py:
```diff
@@ -329,7 +329,7 @@ def _single_tensor_adam(
     has_complex: bool,
-    beta1: float,
+    beta1: Union[float, Tensor],
     beta2: float,
```
missed one here

janeyx99 commented on torch/testing/_internal/common_optimizers.py:
```diff
@@ -507,8 +507,6 @@ def optim_error_inputs_func_adagrad(device, dtype):
 
 
-# TODO: consider tensor LR! See multi_tensor_optimizer_configs in test_optim.py --> tensor LR should work
-# with all implementation code paths...
```
don't delete this todo as tensor LR still doesn't work with noncapturable foreach which is not your problem, but still a problem after this PR

janeyx99 commented on torch/optim/adam.py:
```diff
@@ -483,6 +497,15 @@ def _multi_tensor_adam(
+        {beta1.device: beta1}
+        if isinstance(beta1, Tensor) and str(beta1.device) != "cpu"
+        else None
+    )
```
How come you don't need a beta2_dict?

mlazos replied:
Beta2 is only used in argument positions that allow the tensor to be on different devices. (ie like a scalar tensor which will automatically be moved)

janeyx99 reviewed (commented):
Did some broadstroke reviews. Overall impl is approvable, lmk when CI is green to take another look

janeyx99 commented on torch/optim/optimizer.py:
```diff
@@ -68,6 +68,16 @@ def __repr__(self) -> str:
 required = _RequiredParameter()
 
 
+def _maybe_copy_beta(beta, device, dtype):
```
btw remember to delete

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

janeyx99 commented on torch/optim/adam.py:
```diff
@@ -63,6 +64,23 @@ def __init__(
+            raise ValueError(
+                "betas[0] as a Tensor is not supported for capturable=False and foreach=True"
+            )
+            if betas[0].numel() != 1:
```
this shouldn't go under the previous if nesting

janeyx99 commented on torch/optim/adam.py:
```diff
@@ -63,6 +64,23 @@ def __init__(
+            raise ValueError(
+                "betas[1] as a Tensor is not supported for capturable=False and foreach=True"
+            )
+            if betas[1].numel() != 1:
```
same here

janeyx99 commented on torch/optim/adam.py:
```diff
@@ -465,6 +507,20 @@ def _multi_tensor_adam(
+        raise ValueError(
+            "beta1 as a Tensor is not supported for capturable=False and foreach=True"
+        )
+        if beta1.numel() != 1:
```
same

janeyx99 commented on torch/optim/adam.py:
```diff
@@ -465,6 +507,20 @@ def _multi_tensor_adam(
+        raise ValueError(
+            "beta2 as a Tensor is not supported for capturable=False and foreach=True"
+        )
+        if beta2.numel() != 1:
```
here too

janeyx99 reviewed (commented):
mostly there, though CI is still not green

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

janeyx99 commented on torch/optim/adamw.py:
```diff
@@ -342,6 +364,17 @@ def _single_tensor_adamw(
+    # Note: ensure type declaration is under conditional check for isinstance
+    # or else torchscript will get cranky about the DeviceDict type.
+    if isinstance(beta1, Tensor):
+        beta1_dict: Optional[DeviceDtypeDict] = {}
```
why not start it off with beta1 already inside?

janeyx99 commented on torch/optim/adamw.py:
```diff
@@ -371,8 +404,22 @@ def _single_tensor_adamw(
+        if beta1_dict is not None:
+            key = (device, dtype)
+            if key not in beta1_dict:
+                beta1_dict[key] = beta1.to(device=device, dtype=dtype, non_blocking=True)  # type: ignore[union-attr]
```
Why dtype as well? Is there a case where beta1 dtype shouldn't stay what the user input'd?

mlazos replied:
I needed dtype here for double. Specifically an op below didn't support upcasting the beta (iirc it was addcdiv). By default, the single elem tensor will be float32. I think the foreach ops upcast so this wasn't needed in the multitensor case.

janeyx99 replied:
Hm in this case, the beta would downcast if the params were bf16 or smaller though, which is not desirable. Moreover, the scalar tensor should not run into casting issues for single tensor lerp_ (lerp should do type promotion correctly).

janeyx99 replied:
Do you have the error for when you don't have the dtype?

mlazos replied:
Yeah let me find the error and we can discuss, if it's actually a type promotion bug I can just file an issue. Currently on a plane tho so I'll do it when I land 😅

mlazos replied:
Yeah it still happens - https://gist.github.com/mlazos/1ec56b8a09682105f9b4c17990aed78d

When I remove the dtype from the key it fails with float64 when dtype of the beta is float.

janeyx99 replied:
yea, upcasting only when dtype is fp64 + filing an issue sounds good to me!

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

mlazos commented:
@pytorchbot merge

mlazos commented:
cc @patrick-botco on its way to landing

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

mlazos commented:
@pytorchbot merge

mlazos commented:
@janeyx99 i actually also need to downcast to bfloat16 for the same reason. I'll update the issue accordingly

mlazos committed:
Update on "Support tensor betas in Adam and AdamW"

Adds support for beta1 and beta2 to be wrapped in tensor for Adam and AdamW.

Fixes https://github.com/pytorch/pytorch/issues/133898




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

mlazos commented:
@pytorchbot merge

mlazos commented:
@pytorchbot merge

mlazos commented:
@pytorchbot merge

mlazos commented:
@pytorchbot merge -f "last job stuck"

patrick-botco commented:
thank you all! @mlazos @janeyx99 @awgu

mlazos commented:
> thank you all! @mlazos @janeyx99 @awgu

Yep! Lmk if you need NAdam and RAdam, or if you like you can submit a PR for those and I can review, but glad to help!
